src/test/coins_tests.cpp

```diff
@@ -589,314 +628,245 @@ static void SetCoinsValue(CAmount value, Coin& coin)
+    } else {
+        BOOST_CHECK_EXCEPTION(add_coin(), std::logic_error, HasReason(std::get<std::string>(expected)));
     }
-
-    BOOST_CHECK_EQUAL(result_value, expected_value);
-    BOOST_CHECK_EQUAL(result_flags, expected_flags);
-}
-
-// Simple wrapper for CheckAddCoinBase function above that loops through
-// different possible base_values, making sure each one gives the same results.
-// This wrapper lets the coins_add test below be shorter and less repetitive,
-// while still verifying that the CoinsViewCache::AddCoin implementation
-// ignores base values.
-template <typename... Args>
-static void CheckAddCoin(Args&&... args)
-{
-    for (const CAmount base_value : {ABSENT, SPENT, VALUE1})
-        CheckAddCoinBase(base_value, std::forward<Args>(args)...);
 }
 
 BOOST_AUTO_TEST_CASE(ccoins_add)
 {
     /* Check AddCoin behavior, requesting a new coin from a cache view,
      * writing a modification to the coin, and then checking the resulting
      * entry in the cache after the modification. Verify behavior with the
-     * AddCoin possible_overwrite argument set to false, and to true.
-     *
-     *           Cache   Write   Result  Cache        Result       possible_overwrite
-     *           Value   Value   Value   Flags        Flags
+     * AddCoin coinbase argument set to false, and to true.
+     *               Base        Write   Cache               Expected              Coinbase
      */
-    CheckAddCoin(ABSENT, VALUE3, VALUE3, NO_ENTRY   , DIRTY|FRESH, false);
-    CheckAddCoin(ABSENT, VALUE3, VALUE3, NO_ENTRY   , DIRTY      , true );
-    CheckAddCoin(SPENT , VALUE3, VALUE3, 0          , DIRTY|FRESH, false);
-    CheckAddCoin(SPENT , VALUE3, VALUE3, 0          , DIRTY      , true );
-    CheckAddCoin(SPENT , VALUE3, VALUE3, FRESH      , DIRTY|FRESH, false);
-    CheckAddCoin(SPENT , VALUE3, VALUE3, FRESH      , DIRTY|FRESH, true );
-    CheckAddCoin(SPENT , VALUE3, VALUE3, DIRTY      , DIRTY      , false);
-    CheckAddCoin(SPENT , VALUE3, VALUE3, DIRTY      , DIRTY      , true );
-    CheckAddCoin(SPENT , VALUE3, VALUE3, DIRTY|FRESH, DIRTY|FRESH, false);
-    CheckAddCoin(SPENT , VALUE3, VALUE3, DIRTY|FRESH, DIRTY|FRESH, true );
-    CheckAddCoin(VALUE2, VALUE3, FAIL  , 0          , NO_ENTRY   , false);
-    CheckAddCoin(VALUE2, VALUE3, VALUE3, 0          , DIRTY      , true );
-    CheckAddCoin(VALUE2, VALUE3, FAIL  , FRESH      , NO_ENTRY   , false);
-    CheckAddCoin(VALUE2, VALUE3, VALUE3, FRESH      , DIRTY|FRESH, true );
-    CheckAddCoin(VALUE2, VALUE3, FAIL  , DIRTY      , NO_ENTRY   , false);
-    CheckAddCoin(VALUE2, VALUE3, VALUE3, DIRTY      , DIRTY      , true );
-    CheckAddCoin(VALUE2, VALUE3, FAIL  , DIRTY|FRESH, NO_ENTRY   , false);
-    CheckAddCoin(VALUE2, VALUE3, VALUE3, DIRTY|FRESH, DIRTY|FRESH, true );
+    for (auto base_value : {ABSENT, SPENT, VALUE1}) {
+        CheckAddCoin(base_value, VALUE3, MISSING,            VALUE3_DIRTY_FRESH,   false);
```

Comment:
Can we just hardcode value3 for write? It's the same for every case here.

Reply:
Wouldn't that be inconsistent?
In every other case, the constant contained both values.

Reply:
I mean, if you look at the `Write` column, each value is `VALUE3`.

Reply:
I thought of that, but since it has to be different from `base_value`, seems weird to assume it that in a different method instead of the call-site.
```
             Base        Write   Cache               Expected              Coinbase
CheckAddCoin(base_value, VALUE3, SPENT_DIRTY,        VALUE3_DIRTY,         false); // VALUE3 + VALUE3_DIRTY -> VALUE3_DIRTY
```
makes more sense than
```
             Base        Cache               Expected              Coinbase
CheckAddCoin(base_value, SPENT_DIRTY,        VALUE3_DIRTY,         false); // SPENT_DIRTY -> VALUE3_DIRTY???
```

But I don't mind renaming `VALUE3` to `VAL3` if you're worried about verbosity.